File: src/disjoint-sets-union-find/3378-lcm-connected-components/typescript/lcm_connected_components.ts
import DisjointSets from "@/disjoint-sets-union-find/utils/disjoint_sets";

export class ConnectedComponentsInLCMGraph {
  private union: DisjointSets; // Union for numbers less than or equal threshold
  private nums: number[];
  private belowThresholdNums: number[] = [];
  private threshold: number;
  private singleComponents: number[] = [];
  constructor(nums: number[], threshold: number) {
    this.nums = nums;
    this.threshold = threshold;
    this.union = new DisjointSets(threshold);
    this.setup();
  }

  private setup(): void {
    for (const num of this.nums) {
      if (num > this.threshold) {
        this.singleComponents.push(num);
        continue;
      }
      this.belowThresholdNums.push(num);
      let m = 1;
      while (num * m <= this.threshold) {
        this.union.unionSets(num, num * m);
        m += 1;
      }
    }
  }
  public countComponents(): number {
    const results = new Set(
      this.belowThresholdNums.map((num) => this.union.findRoot(num)),
    );
    return results.size + this.singleComponents.length;
  }
}

File: src/disjoint-sets-union-find/3378-lcm-connected-components/cpp/src/lcm_connected_components.cpp
#include "lcm_connected_components.hpp"

#include <unordered_set>

ConnectedComponentsInLCMGraph::ConnectedComponentsInLCMGraph(
    std::vector<int> nums, int threshold)
    : _nums(std::move(nums)),
      _threshold(threshold),
      _disjointSets(threshold)
{
  setup();
}

void ConnectedComponentsInLCMGraph::setup()
{
  for (int num : _nums)
  {
    if (num > _threshold)
    {
      _singleComponents.push_back(num);
      continue;
    }

    _belowThresholdNums.push_back(num);
    long long multiple = num;
    while (multiple <= _threshold)
    {
      _disjointSets.unite(num, static_cast<int>(multiple));
      multiple += num;
    }
  }
}

int ConnectedComponentsInLCMGraph::countComponents()
{
  std::unordered_set<int> roots;
  for (int num : _belowThresholdNums)
  {
    roots.insert(_disjointSets.find(num));
  }
  return static_cast<int>(roots.size() + _singleComponents.size());
}
